File: src/main/java/org/radarcns/monitor/DisconnectMonitor.java
/*
 * Copyright 2017 King's College London and The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.monitor;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.text.DateFormat;
import java.text.Format;
import java.time.Duration;
import java.time.Instant;
import java.util.Collection;
import java.util.Date;
import java.util.Iterator;
import java.util.Locale;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import javax.mail.MessagingException;
import org.apache.avro.generic.GenericRecord;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.radarcns.config.monitor.DisconnectMonitorConfig;
import org.radarcns.config.RadarPropertyHandler;
import org.radarcns.kafka.ObservationKey;
import org.radarcns.monitor.DisconnectMonitor.DisconnectMonitorState;
import org.radarcns.util.EmailSender;
import org.radarcns.util.EmailSenders;
import org.radarcns.util.Monitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Monitors whether an ID has stopped sending measurements and sends an email when this occurs.
 */
public class DisconnectMonitor extends AbstractKafkaMonitor<
        GenericRecord, GenericRecord, DisconnectMonitorState> {

    private static final Logger logger = LoggerFactory.getLogger(DisconnectMonitor.class);

    private final ScheduledExecutorService scheduler;
    private final Duration timeUntilReportedMissing;
    private final EmailSenders senders;
    private final Format dayFormat;
    private final int numRepetitions;
    private final Duration repeatInterval;
    private final Duration minRepetitionInterval;
    private final Monitor monitor;
    private final String message;

    public DisconnectMonitor(RadarPropertyHandler radar, Collection<String> topics, String groupId,
                             EmailSenders senders) {
        super(radar, topics, groupId, "1", new DisconnectMonitorState());
        this.senders = senders;
        this.dayFormat = DateFormat.getDateTimeInstance(
                DateFormat.MEDIUM, DateFormat.SHORT, Locale.US);
        this.scheduler = Executors.newSingleThreadScheduledExecutor();
        this.monitor = new Monitor(logger, " records monitored for Disconnect");

        DisconnectMonitorConfig config = radar.getRadarProperties().getDisconnectMonitor();
        this.timeUntilReportedMissing = Duration.ofSeconds(config.getTimeout());
        this.numRepetitions = config.getAlertRepetitions();
        this.repeatInterval = Duration.ofSeconds(config.getAlertRepeatInterval());
        this.message = config.getMessage();
        this.minRepetitionInterval = repeatInterval.dividedBy(10);

        Properties props = new Properties();
        props.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest");
        configure(props);

        super.setPollTimeout(timeUntilReportedMissing);
    }

    /**
     * Schedules the repetitive alert task.
     */
    @Override
    public void start() {
        scheduler.scheduleAtFixedRate(this.monitor, 5, 5, TimeUnit.MINUTES);
        startScheduler();
        super.start();
    }

    @Override
    public void shutdown() {
        super.shutdown();
        scheduler.shutdown();
    }

    /**
     * Starts the scheduled alert updates.
     * Protected method to support unit testing
     */
    protected void startScheduler() {
        if (numRepetitions > 0) {
            logger.info("Start scheduled alert updates with the delay of {}", repeatInterval);
            state.reportedMissing.forEach(this::scheduleRepetition);
        }
    }

    @Override
    protected void evaluateRecords(ConsumerRecords<GenericRecord, GenericRecord> records) {
        super.evaluateRecords(records);

        Instant reportThreshold = Instant.now().minus(timeUntilReportedMissing);

        Iterator<Map.Entry<String, Long>> iterator = state.lastSeen.entrySet().iterator();

        while (iterator.hasNext()) {
            Map.Entry<String, Long> entry = iterator.next();
            // calculate timeout from current timestamp per device
            long lastSeen = entry.getValue();
            if (reportThreshold.isAfter(Instant.ofEpochMilli(lastSeen))) {
                String missingKey = entry.getKey();
                // remove processed records to prevent adding alerts multiple times
                iterator.remove();
                reportMissing(missingKey, new MissingRecordsReport(lastSeen));
            }
        }
    }

    @Override
    protected void evaluateRecord(ConsumerRecord<GenericRecord, GenericRecord> record) {
        ObservationKey key = extractKey(record);

        this.monitor.increment();

        long now = System.currentTimeMillis();
        String keyString = getStateStore().keyToString(key);
        state.lastSeen.put(keyString, now);

        MissingRecordsReport missingReport = state.reportedMissing.remove(keyString);
        if (missingReport != null) {
            missingReport.cancelRepetition();
            reportRecovered(key, missingReport.getReportedMissing());
        }
    }

    /**
     * Schedule a missing device message to be sent again.
     * @param key record key
     * @param report missing record details
     */
    private void scheduleRepetition(final String key, final MissingRecordsReport report) {
        if (report.getMessageNumber() < numRepetitions) {
            long reportedMissing = report.getReportedMissing();
            Instant now = Instant.now();
            Duration passedInterval = Duration.between(Instant.ofEpochMilli(reportedMissing), now);

            Duration nextRepetition;
            if (minRepetitionInterval.compareTo(repeatInterval.minus(passedInterval)) >= 0) {
                nextRepetition = minRepetitionInterval;
            } else {
                nextRepetition = repeatInterval.minus(passedInterval);
            }

            report.setFuture(scheduler.schedule(() -> reportMissing(key, report.newRepetition()),
                    nextRepetition.toMillis(), TimeUnit.MILLISECONDS));
        }
    }

    private void reportMissing(String keyString, MissingRecordsReport report) {
        ObservationKey key = getStateStore().stringToKey(keyString);

        // Don't report if no email address for this projectId
        EmailSender sender = senders.getEmailSenderForProject(key.getProjectId());
        if (sender == null) {
            return;
        }

        long timeout = report.getTimeout();
        logger.info("Device {} timeout {} (message {} of {}). Reporting it missing.", key,
                timeout, report.getMessageNumber(), numRepetitions);
        try {
            String lastSeen = dayFormat.format(report.getLastSeenDate());
            StringBuilder textBuilder = new StringBuilder(400);
            StringBuilder subjectBuilder = new StringBuilder(100);
            textBuilder.append("The device ")
                    .append(key)
                    .append(" seems disconnected. It was last seen on ")
                    .append(lastSeen)
                    .append(" (")
                    .append(String.valueOf(timeout / 1000L))
                    .append(" seconds ago). If this is not intended, please ensure that it gets reconnected.");
            if (message != null) {
                textBuilder.append("\n\n").append(message);
            }
            subjectBuilder.append("[RADAR] Device has disconnected");
            if (numRepetitions > 0 && report.getMessageNumber() == numRepetitions) {
                textBuilder.append("\n\nThis is the final warning email for this device.");
                subjectBuilder.append(". Final message");
            } else if (numRepetitions > 0) {
                textBuilder.append("\n\nThis is warning number ")
                        .append(String.valueOf(report.getMessageNumber()))
                        .append(" of ")
                        .append(String.valueOf(numRepetitions));
            }

            sender.sendEmail(subjectBuilder.toString(), textBuilder.toString());
            logger.debug("Sent disconnected message successfully");
        } catch (MessagingException mex) {
            logger.error("Failed to send disconnected message.", mex);
        } finally {
            // store last seen and reportedMissing timestamp
            state.getReportedMissing().put(keyString, report);
            scheduleRepetition(keyString, report);
        }
    }

    private void reportRecovered(ObservationKey key, long reportedMissingTime) {
        // Don't report if no email address for this projectId
        EmailSender sender = senders.getEmailSenderForProject(key.getProjectId());
        if (sender == null) {
            return;
        }

        logger.info("Device {} seen again. Reporting it recovered.", key);
        try {
            Date reportedMissingDate = new Date(reportedMissingTime);
            String reportedMissing = dayFormat.format(reportedMissingDate);

            sender.sendEmail("[RADAR] device has reconnected",
                    "The device " + key + " that was reported disconnected on "
                            + reportedMissing + " has reconnected: it is sending new data.");
            logger.debug("Sent reconnected message successfully");
        } catch (MessagingException mex) {
            logger.error("Failed to send reconnected message.", mex);
        }
    }

    /**
     * State of disconnect monitor.
     */
    public static class DisconnectMonitorState {
        private final Map<String, Long> lastSeen = new ConcurrentHashMap<>();
        private final Map<String, MissingRecordsReport> reportedMissing = new ConcurrentHashMap<>();

        public Map<String, Long> getLastSeen() {
            return lastSeen;
        }

        public void setLastSeen(Map<String, Long> lastSeen) {
            this.lastSeen.putAll(lastSeen);
        }

        public Map<String, MissingRecordsReport> getReportedMissing() {
            return reportedMissing;
        }

        public void setReportedMissing(Map<String, MissingRecordsReport> reportedMissing) {
            this.reportedMissing.putAll(reportedMissing);
        }
    }

    /**
     * Stores data of data from missing records alert
     * such as lastSeen and reportedTime.
     */
    public static class MissingRecordsReport {
        private final long lastSeen;
        private final long reportedMissing;
        private final int messageNumber;

        @JsonIgnore
        private Future<?> future;

        @JsonCreator
        public MissingRecordsReport(@JsonProperty("lastSeen") long lastSeen,
                @JsonProperty("reportedMissing") long reportedMissing,
                @JsonProperty("messagesProcessed") int messageNumber) {
            this.lastSeen = lastSeen;
            this.reportedMissing = reportedMissing;
            this.messageNumber = messageNumber;
            this.future = null;
        }

        public MissingRecordsReport(long lastSeen) {
            this(lastSeen, System.currentTimeMillis(), 0);
        }

        public long getLastSeen() {
            return lastSeen;
        }

        public long getReportedMissing() {
            return reportedMissing;
        }

        @JsonIgnore
        public long getTimeout() {
            return reportedMissing - lastSeen;
        }

        @JsonIgnore
        public Date getLastSeenDate() {
            return new Date(lastSeen);
        }

        public int getMessageNumber() {
            return messageNumber;
        }

        public MissingRecordsReport newRepetition() {
            return new MissingRecordsReport(
                    lastSeen, System.currentTimeMillis(), messageNumber + 1);
        }

        public synchronized void cancelRepetition() {
            if (future != null) {
                future.cancel(true);
            }
        }

        public synchronized void setFuture(Future<?> future) {
            this.future = future;
        }
    }
}

File: src/main/kotlin/org/radarbase/kotlin/monitor/DisconnectMonitor.kt
/*
 * Copyright 2017 King's College London and The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.kotlin.monitor

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import org.apache.avro.generic.GenericRecord
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.consumer.ConsumerRecords
import org.radarbase.kotlin.config.RadarPropertyHandler
import org.radarcns.kafka.ObservationKey
import org.radarbase.util.EmailSenders
import org.radarbase.kotlin.util.Monitor
import org.slf4j.LoggerFactory
import java.text.DateFormat
import java.time.Duration
import java.time.Instant
import java.util.*
import java.util.concurrent.*
import javax.mail.MessagingException

/**
 * Monitors whether an ID has stopped sending measurements and sends an email when this occurs.
 */
class DisconnectMonitor(
    radar: RadarPropertyHandler,
    topics: Collection<String>,
    groupId: String,
    private val senders: EmailSenders?
) : AbstractKafkaMonitor<GenericRecord, GenericRecord, DisconnectMonitor.DisconnectMonitorState>(
    radar, topics, groupId, "1", DisconnectMonitorState()
) {
    private val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private val timeUntilReportedMissing: Duration
    private val dayFormat: DateFormat = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.SHORT, Locale.US)
    private val numRepetitions: Int
    private val repeatInterval: Duration
    private val minRepetitionInterval: Duration
    private val monitor: Monitor = Monitor(logger, " records monitored for Disconnect")
    private val message: String?

    init {
        val config = radar.radarProperties.disconnectMonitor
        timeUntilReportedMissing = Duration.ofSeconds(config!!.timeout)
        numRepetitions = config.alertRepetitions
        repeatInterval = Duration.ofSeconds(config.alertRepeatInterval)
        message = config.message
        minRepetitionInterval = repeatInterval.dividedBy(10)

        val props = Properties()
        props.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
        configure(props)

        pollTimeout = timeUntilReportedMissing
    }

    override fun start() {
        scheduler.scheduleAtFixedRate(monitor as Runnable, 5, 5, TimeUnit.MINUTES)
        startScheduler()
        super.start()
    }

    override fun shutdown() {
        super.shutdown()
        scheduler.shutdown()
    }

    internal fun startScheduler() {
        if (numRepetitions > 0) {
            logger.info("Start scheduled alert updates with the delay of {}", repeatInterval)
            state?.reportedMissing?.forEach { (key, report) -> scheduleRepetition(key, report) }
        }
    }

    override fun evaluateRecords(records: ConsumerRecords<GenericRecord, GenericRecord>) {
        super.evaluateRecords(records)

        val reportThreshold = Instant.now().minus(timeUntilReportedMissing)
        val iterator = state?.lastSeen?.entries?.iterator() ?: return

        while (iterator.hasNext()) {
            val entry = iterator.next()
            val lastSeen = entry.value
            if (reportThreshold.isAfter(Instant.ofEpochMilli(lastSeen))) {
                val missingKey = entry.key
                iterator.remove()
                reportMissing(missingKey, MissingRecordsReport(lastSeen))
            }
        }
    }

    override fun evaluateRecord(record: ConsumerRecord<GenericRecord, GenericRecord>) {
        val key = extractKey(record)
        monitor.increment()

        val now = System.currentTimeMillis()
        val keyString = getStateStore()!!.keyToString(key)
        state?.lastSeen?.put(keyString, now)

        val missingReport = state?.reportedMissing?.remove(keyString)
        if (missingReport != null) {
            missingReport.cancelRepetition()
            reportRecovered(key, missingReport.reportedMissing)
        }
    }

    private fun scheduleRepetition(key: String, report: MissingRecordsReport) {
        if (report.messageNumber < numRepetitions) {
            val reportedMissing = report.reportedMissing
            val now = Instant.now()
            val passedInterval = Duration.between(Instant.ofEpochMilli(reportedMissing), now)

            val nextRepetition = if (minRepetitionInterval >= repeatInterval.minus(passedInterval)) {
                minRepetitionInterval
            } else {
                repeatInterval.minus(passedInterval)
            }

            report.future = scheduler.schedule({ reportMissing(key, report.newRepetition()) },
                nextRepetition.toMillis(), TimeUnit.MILLISECONDS)
        }
    }

    private fun reportMissing(keyString: String, report: MissingRecordsReport) {
        val key = getStateStore()!!.stringToKey(keyString)
        val sender = senders?.getEmailSenderForProject(key.projectId) ?: return

        val timeout = report.timeout
        logger.info("Device {} timeout {} (message {} of {}). Reporting it missing.", key,
            timeout, report.messageNumber, numRepetitions)

        try {
            val lastSeen = dayFormat.format(report.lastSeenDate)
            var text = "The device $key seems disconnected. It was last seen on $lastSeen (${timeout / 1000L} seconds ago). If this is not intended, please ensure that it gets reconnected."
            if (message != null) {
                text += "\n\n$message"
            }
            var subject = "[RADAR] Device has disconnected"
            if (numRepetitions > 0 && report.messageNumber == numRepetitions) {
                text += "\n\nThis is the final warning email for this device."
                subject += ". Final message"
            } else if (numRepetitions > 0) {
                text += "\n\nThis is warning number ${report.messageNumber} of $numRepetitions"
            }

            sender.sendEmail(subject, text)
            logger.debug("Sent disconnected message successfully")
        } catch (mex: MessagingException) {
            logger.error("Failed to send disconnected message.", mex)
        } finally {
            state?.reportedMissing?.put(keyString, report)
            scheduleRepetition(keyString, report)
        }
    }

    private fun reportRecovered(key: ObservationKey, reportedMissingTime: Long) {
        val sender = senders?.getEmailSenderForProject(key.projectId) ?: return

        logger.info("Device {} seen again. Reporting it recovered.", key)
        try {
            val reportedMissingDate = Date(reportedMissingTime)
            val reportedMissing = dayFormat.format(reportedMissingDate)

            sender.sendEmail("[RADAR] device has reconnected",
                "The device $key that was reported disconnected on $reportedMissing has reconnected: it is sending new data.")
            logger.debug("Sent reconnected message successfully")
        } catch (mex: MessagingException) {
            logger.error("Failed to send reconnected message.", mex)
        }
    }

    class DisconnectMonitorState {
        val lastSeen: MutableMap<String, Long> = ConcurrentHashMap()
        val reportedMissing: MutableMap<String, MissingRecordsReport> = ConcurrentHashMap()

        fun setLastSeen(lastSeen: Map<String, Long>) {
            this.lastSeen.putAll(lastSeen)
        }

        fun setReportedMissing(reportedMissing: Map<String, MissingRecordsReport>) {
            this.reportedMissing.putAll(reportedMissing)
        }
    }

    class MissingRecordsReport @JsonCreator constructor(
        @JsonProperty("lastSeen") val lastSeen: Long,
        @JsonProperty("reportedMissing") val reportedMissing: Long,
        @JsonProperty("messageNumber") val messageNumber: Int
    ) {
        @JsonIgnore
        @Volatile
        var future: Future<*>? = null

        constructor(lastSeen: Long) : this(lastSeen, System.currentTimeMillis(), 0)

        @get:JsonIgnore
        val timeout: Long
            get() = reportedMissing - lastSeen

        @get:JsonIgnore
        val lastSeenDate: Date
            get() = Date(lastSeen)

        fun newRepetition(): MissingRecordsReport = MissingRecordsReport(lastSeen, System.currentTimeMillis(), messageNumber + 1)

        fun cancelRepetition() {
            future?.cancel(true)
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DisconnectMonitor::class.java)
    }
}
